Lib/traceback.py

```diff
@@ -621,3 +624,83 @@ def format(self, *, chain=True):
         yield from self.stack.format()
         yield from self.format_exception_only()
+
+
+class TracebackExceptionGroup:
+    """An exception or exception group ready for rendering.
+
+    The traceback module captures enough attributes from the original exception
+    group to this intermediary form to ensure that no references are held, while
+    still being able to fully print or format it.
+
+    Use `from_exception` to create TracebackExceptionGroup instances from exception
+    objects, or the constructor to create TracebackExceptionGroup instances from
+    individual components.
+
+    - :attr:`summary` A tuple of (indentation level, TracebackException) pairs
+    """
+
+    def __init__(self, exc_type, exc_value, exc_traceback, **kwargs):
+        self.summary = tuple(
+            self._gen_summary(
+                exc_type, exc_value, exc_traceback, **kwargs))
+        self._str = _some_str(exc_value)
+
+    def _gen_summary(self, exc_type, exc_value, exc_traceback,
+                     indent=0, **kwargs):
+        # recursively extract the sequence of (indent, TracebackException)
+        # pairs. If exc_value is a single exception, there is only one pair
+        # (with indent 0). If it is an exception group, we get one pair
+        # for each exception in it, showing the whole tree
+        te = TracebackException(
+            exc_type, exc_value, exc_traceback, **kwargs)
+        yield tuple((indent, te))
+        if isinstance(exc_value, exception_group.ExceptionGroup):
+            for e in exc_value.excs:
+                yield from self._gen_summary(
+                    type(e), e, e.__traceback__, indent=indent+4, **kwargs)
+
+    @classmethod
+    def from_exception(cls, exc, *args, **kwargs):
+        """Create a TracebackException from an exception."""
+        return cls(type(exc), exc, exc.__traceback__, *args, **kwargs)
+
+    def format(self, *, chain=True):
+        """Format the exception or exception group.
+
+        For an exception group, the shared part of the traceback is printed,
+        followed by a printout of each exception in the group, which is
+        expanded recursively.
+
+        If chain is not *True*, *__cause__* and *__context__* will not be formatted.
+
+        The return value is a generator of strings, each ending in a newline and
+        some containing internal newlines. `print_exception_group` is a wrapper
+        around this method which just prints the lines to a file.
+        """
+        # TODO: should we add an arg to bound the number of exceptions printed?
+        # or should we use limit somehow for that?
+        for indent, te in self.summary:
+            if indent == 0:
+                yield from te.format(chain=chain)
+            else:
+                indent_str = ' '*indent
+                yield indent_str + '-'*(60-indent) +'\n'
+                yield textwrap.indent(
+                    ''.join(list(te.format(chain=chain))), indent_str)
```

Comment:
I worry that this causes long blocks with internal newlines. Perhaps it would be better to do something like
```py
for line in te.format(chain=chain):
    yield indent_str + line
```
except that if line contains an internal newline we'd have to do the textwrap.indent() thing.

Do you know under what circumstances te.format() returns lines with internal newlines? It's not common, I think.

Reply:
From scanning the code the only place I see internal newlines is in StackSummary.format:

```
            if frame.locals:
                for name, value in sorted(frame.locals.items()):
                    row.append('    {name} = {value}\n'.format(name=name, value=value))
            result.append(''.join(row))
```

We could add a "prefix" arg to all the format* functions in this module, and then the indent happens when each line is created.

Reply:
Example:

```
def f():
    x,y,z = 1,2,3
    (x+y+z)/0

try:
    f()
except Exception as e:
    import traceback
    print(repr(list(traceback.TracebackException.from_exception(e, capture_locals=True).format())[-2]))

```

Output is:
`'  File "C:\\Users\\User\\src\\cpython\\x.py", line 4, in f\n    (x+y+z)/0\n    x = 1\n    y = 2\n    z = 3\n'`

Reply:
Ugh, I worry that someone, somewhere depends on the local variable summary being attached to the "File ..." string.

So I propose something like this:
```py
for line in te.format(chain=chain):
    if '\n' in line:
        yield indent_str + line
    else:
        yield <something using textwrap.indent>
```
(Even though it might be a premature optimization. :-)